### PR TITLE
Issue #13809: Kill mutation in CommonUtil

### PR DESCRIPTION
**Pull request by Kevin222004**
Issue #13809: Kill mutation in CommonUtil

regression is taken from https://github.com/checkstyle/checkstyle/pull/13284
gist: https://gist.github.com/romani/f15f3afabac0f9ad1e63f429b62d445f

Diff Regression config: https://gist.githubusercontent.com/romani/f15f3afabac0f9ad1e63f429b62d445f/raw/40dfb3e50dc3e5606844f00d074f572be4682079/Config.xml
Diff Regression projects: https://gist.githubusercontent.com/Kevin222004/9600f179b602d4c971bdb0a050099005/raw/360a95ed7bb60d7a0956e531199d484c4d6f6617/test-projects.properties
Report label: Regression-with-path-for-basedir

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java [hunk 1]
```diff
@@ -261,8 +261,8 @@ public static String relativizeAndNormalizePath(final String baseDirectory, fina
             resultPath = path;
         }
         else {
-            final Path pathAbsolute = Paths.get(path).normalize();
-            final Path pathBase = Paths.get(baseDirectory).normalize();
+            final Path pathAbsolute = Paths.get(path);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java:264]: Looks like we need to test message of violation and validate how path is starting https://www.google.com/amp/s/www.geeksforgeeks.org/path-normalize-method-in-java-with-examples/amp/

To be OS agnostic we usually avoid checking full path in violation, we might do only file name

**Review by romani — CHANGES_REQUESTED** [review]
Items :

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java (hunk 1 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java:264]: What we can test is ... We set path with `.` and/or `..` but have in violation path without such symbols.

Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java:264]: I have tried multiple ways but at the i didnt find a way in which relativize path is getting affected.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java:264]: yes, such normalization is done by jdk
https://github.com/openjdk/jdk/blob/db3402577a2c14a41045753a1ffe2829a6bdda91/src/java.base/unix/classes/sun/nio/fs/UnixPath.java#L478-L495

we already have test on this https://github.com/checkstyle/checkstyle/blob/bd3a9bb9ec9cf804f628171c2e7557b0afbcfd93/src/test/java/com/puppycrawl/tools/checkstyle/utils/CommonUtilTest.java#L221-L224

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java:264]: image

**Review by romani — APPROVED** [review]
ok to merge

**Comment by rnveach** [comment]
1) Where is regression showing this won't be an issue? This is used by Checker.
2) This method is not really called "relativizeAnd**Normalize**Path" if there is no more normalize.

**Comment by romani** [comment]
> Where is regression showing this won't be an issue? This is used by Checker.

regression is only covering behavior on config, to make this regression something weird should be done with paths to validate. non of our tools can do this.

> This method is not really called "relativizeAndNormalizePath" if there is no more normalize.

relativize does normalize. Do you propose to change name of util method ?

**Comment by rnveach** [comment]
Regression should also cover file paths which is done by Checker, as 2 paths need to line up for a match for differencing. If it truthfully has no impact, regression will show no differences either way and will show we won't have anything to worry about after merge.

I am leaning to changing the util name and javadoc since I assume it was named normalize for the normalize that happened in it.

**Comment by romani** [comment]
method is renamed.

**Comment by romani** [comment]
Github, generate report

**Comment by romani** [comment]
@rnveach , please review.

**Comment by rnveach** [comment]
@romani Please make sure you are using a config to hit the code. `baseDir` has to be set to trigger as shown in the method in this pr.

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java (hunk 1 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java:264]: > We do not need to care about path in zips 

You may want to recheck this. Some users put configuration files into external JARs controlled by Maven. Those are basically ZIP files.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java:264]: We will happily fix issue if they will show how to reproduce issue.

**Comment by romani** [comment]
Github, generate repor

**Comment by romani** [comment]
Github, generate report

**Review by rdiachenko — APPROVED** [review]
lgtm

normalization is done as part of `relativize` for UnixPath

**Comment by romani** [comment]
regression report needs to pass, I am not sure why it is failing with exception.

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Vyom-Yadav** [comment]
@romani @Kevin222004 I think there is a problem with relativizing path.
```
Caused by: java.lang.IllegalArgumentException: 'other' is different type of Path
    at sun.nio.fs.UnixPath.relativize (UnixPath.java:429)
    at sun.nio.fs.UnixPath.relativize (UnixPath.java:43)
    at com.puppycrawl.tools.checkstyle.utils.CommonUtil.relativizeAndNormalizePath (CommonUtil.java:266)
```
You may want to take a look at https://stackoverflow.com/q/16299604/15412365

**Comment by romani** [comment]
problem is at `relativizeAndNormalizePath` , so it is execution without update of this PR.
so exception is not related to PR update, and this is just bad config.

**Comment by romani** [comment]
Github, generate report

**Comment by romani** [comment]
Github, generate report

**Comment by romani** [comment]
@rnveach , report is ready.

**Review by romani — APPROVED** [review]
Ok to merge

**Comment by rnveach** [comment]
@romani So we are not going to do anything on https://github.com/checkstyle/checkstyle/pull/13953#discussion_r1375561273 ?

**Comment by romani** [comment]
@rnveach , yes, let's have regression if it's actually matters.
I don't know how to deal with such path in zip for checkstyle.

**Review by rnveach — APPROVED** [review]

**Review by nrmancuso — APPROVED** [review]

**Comment by rnveach** [comment]
Apparently this change failed in backport (JDK 8). Nothing changed in the javadoc for `Path.relativize`.

**Pull request merged**